### PR TITLE
Refactoring to prepare for version 15

### PR DESCRIPTION
**Pull request by jeff-davis**
Allows compilation against PG15. Work remains to port the ruleutils changes, ensure tests pass, and support v15 in CI.

Did not update `configure` script yet.

### CODE REVIEW

**Comment by jeff-davis** [comment]
I split into individual commits for ease-of-review. 

This doesn't offer full PG15 support yet, but it gets pretty close.

File: src/backend/columnar/columnar_storage.c [hunk 1]
```diff
@@ -354,8 +356,7 @@ ColumnarStorageGetReservedOffset(Relation rel, bool force)
 bool
 ColumnarStorageIsCurrent(Relation rel)
 {
-	RelationOpenSmgr(rel);
-	BlockNumber nblocks = smgrnblocks(rel->rd_smgr, MAIN_FORKNUM);
+	BlockNumber nblocks = smgrnblocks(RelationGetSmgr(rel), MAIN_FORKNUM);
```
Comment (anarazel) [inline src/backend/columnar/columnar_storage.c:359]: These are problematic before 15, FWIW.

Comment (jeff-davis) [reply src/backend/columnar/columnar_storage.c:359]: Hmm, I thought if you opened right before using the `rel->rd_smgr`, it was fine. To be a problem doesn't there need to be some other call in between the `RelationOpenSmgr()` and the `smgr*()` call that uses it?

**Review by marcocitus — APPROVED** [review]

**Comment by onderkalaci** [comment]
Hmm, shouldn't we need to adjust configure script? When I tried something on PG 15, I ended up doing: [WIP: Detect connection socket close by onderkalaci · Pull Request #5908 · citusdata/citus (github.com)](https://github.com/citusdata/citus/pull/5908/commits/4f330f8d49f8d8d878414794ed57e6170a7cda32#diff-90d08e583c4c9c6f391b2ae90f819f600a6326928ea9512c9e0c6d98e9f29ac2)



Also, I presume this PR does not cover all Citus related things, because we should be preparing the ruleutils? At very least, I tried enabling it as follows: [WIP: Detect connection socket close by onderkalaci · Pull Request #5908 · citusdata/citus (github.com)](https://github.com/citusdata/citus/pull/5908/files#diff-b5ee86823b5a5743dc453b01f49cda17a8ee9e34c80089b07513252c3221dfd0).   Might be useful to document what we are achieving on the PR description. 

Thanks, it is great to start testing Citus with PG 15!

**Comment by marcocitus** [comment]
I read this as a refactor to prepare for PG15, rather than PG15 compatibility (updated the PR title a bit). Would be good to update the PR description with the changes.

**Pull request merged**